src/hotspot/cpu/aarch64/universalNativeInvoker_aarch64.cpp

```diff
@@ -125,3 +129,254 @@ address ProgrammableInvoker::generate_adapter(jobject jabi, jobject jlayout) {
+  OopMapSet* _oop_maps;
+public:
+  NativeInvokerGenerator(CodeBuffer* buffer,
+                         BasicType* signature,
+                         int num_args,
+                         BasicType ret_bt,
+                         int shadow_space_bytes,
+                         const GrowableArray<VMReg>& input_registers,
+                         const GrowableArray<VMReg>& output_registers)
+   : StubCodeGenerator(buffer, PrintMethodHandleStubs),
+     _signature(signature),
+     _num_args(num_args),
+     _ret_bt(ret_bt),
+     _shadow_space_bytes(shadow_space_bytes),
+     _input_registers(input_registers),
+     _output_registers(output_registers),
+     _frame_complete(0),
+     _framesize(0),
+     _oop_maps(NULL) {
+    assert(_output_registers.length() <= 1
+           || (_output_registers.length() == 2 && !_output_registers.at(1)->is_valid()), "no multi-reg returns");
+  }
+
+  void generate();
+
+  int frame_complete() const {
+    return _frame_complete;
+  }
+
+  int framesize() const {
+    return (_framesize >> (LogBytesPerWord - LogBytesPerInt));
+  }
+
+  OopMapSet* oop_maps() const {
+    return _oop_maps;
+  }
+
+private:
+#ifdef ASSERT
+  bool target_uses_register(VMReg reg) {
+    return _input_registers.contains(reg) || _output_registers.contains(reg);
+  }
+#endif
+};
+
+static const int native_invoker_code_size = 1024;
+
+RuntimeStub* ProgrammableInvoker::make_native_invoker(BasicType* signature,
+                                                      int num_args,
+                                                      BasicType ret_bt,
+                                                      int shadow_space_bytes,
+                                                      const GrowableArray<VMReg>& input_registers,
+                                                      const GrowableArray<VMReg>& output_registers) {
+  int locs_size  = 64;
+  CodeBuffer code("nep_invoker_blob", native_invoker_code_size, locs_size);
+  NativeInvokerGenerator g(&code, signature, num_args, ret_bt, shadow_space_bytes, input_registers, output_registers);
+  g.generate();
+  code.log_section_sizes("nep_invoker_blob");
+
+  RuntimeStub* stub =
+    RuntimeStub::new_runtime_stub("nep_invoker_blob",
+                                  &code,
+                                  g.frame_complete(),
+                                  g.framesize(),
+                                  g.oop_maps(), false);
+
+  if (TraceNativeInvokers) {
+    stub->print_on(tty);
+  }
+
+  return stub;
+}
+
+void NativeInvokerGenerator::generate() {
+  // we can't use rscratch1 because it is r8, and used by the ABI
+  Register scratch1 = r9;
+  Register scratch2 = r10;
+  assert(!target_uses_register(scratch1->as_VMReg()), "conflict");
+  assert(!target_uses_register(scratch2->as_VMReg()), "conflict");
+  assert(!target_uses_register(rthread->as_VMReg()), "conflict");
+
+  enum layout {
+    rbp_off,
+    rbp_off2,
+    return_off,
+    return_off2,
```

Comment:
It's a bit weird that we use the x86 register names here (although sharedRuntime_aarch64.cpp has this already). `rfp_off` and `lr_off` would be more accurate.